Watcher sometimes recording old state

This is a hard to reproduce error, as it occurs only sporadically. I have not been able to pin down the exact circumstances under which it occurs.

As far as I have been able to understand, the wrong behavior is as follows:
1. I use `zodbsync pick` to pick some commits, which modifies some objects as well as adding some. The changed or added objects are played back into the ZODB.
2. The watcher notices that the objects have been changed and starts to record them
3. Of the changed objects, the watcher records their *old* state instead of the new one, overwriting the state that was created by the pick.

The correct new state of the objects is present in Zope, but not on the file system. Further sweeps do not pick up any additional transactions and do not correct this.

This should not happen. The watcher should always see the objects in the state after the transaction currently stored in `self.last_visible_txn`.

This might be related to the watcher tests failing sporadically, which is also hard to pin down. One example traceback is:
```
===================================================== FAILURES ======================================================
____________________________________________ TestSync.test_watch_change _____________________________________________

self = <perfact.zodbsync.tests.test_sync.TestSync object at 0x7fa274097e80>
conn = <perfact.zodbsync.helpers.Namespace object at 0x7fa2740d1eb0>

    def test_watch_change(self, conn):
        """
        Start the watcher, change something using the second connection without
        commiting yet, do a step on the watcher, make sure the change is not
        yet visible, then commit the change and do another step, making sure
        that it is now present.
        """
        fname = self.repo.path + '/__root__/__meta__'
        watcher = self.runner('watch')
        watcher.setup()
        conn.tm.begin()
        conn.app._addRole('TestRole')
        watcher.step()
        assert 'TestRole' not in open(fname).read()
        conn.tm.commit()
        watcher.step()
>       assert 'TestRole' in open(fname).read()
E       assert 'TestRole' in "[\n    ('is_root', True),\n    ('perms', [\n        ('Access contents information', False, ['Anonymous', 'Manager']),...bDAV access', False, ['Authenticated', 'Manager']),\n        ]),\n    ('title', 'Zope'),\n    ('type', 'Folder'),\n]\n"
E        +  where "[\n    ('is_root', True),\n    ('perms', [\n        ('Access contents information', False, ['Anonymous', 'Manager']),...bDAV access', False, ['Authenticated', 'Manager']),\n        ]),\n    ('title', 'Zope'),\n    ('type', 'Folder'),\n]\n" = <built-in method read of _io.TextIOWrapper object at 0x7fa2740e7380>()
E        +    where <built-in method read of _io.TextIOWrapper object at 0x7fa2740e7380> = <_io.TextIOWrapper name='/tmp/tmp0ox63v_h/__root__/__meta__' mode='r' encoding='UTF-8'>.read
E        +      where <_io.TextIOWrapper name='/tmp/tmp0ox63v_h/__root__/__meta__' mode='r' encoding='UTF-8'> = open('/tmp/tmp0ox63v_h/__root__/__meta__')

perfact/zodbsync/tests/test_sync.py:356: AssertionError
```

Another error in the tests than sometimes happens:
```
===================================================== FAILURES ======================================================
_____________________________________________ TestSync.test_watch_move ______________________________________________

self = <perfact.zodbsync.tests.test_sync.TestSync object at 0x7f66180bcbe0>
conn = <perfact.zodbsync.helpers.Namespace object at 0x7f66180a3dc0>

    def test_watch_move(self, conn):
        """
        Create a Page Template, record it using the watcher, rename it and make
        sure the watcher notices. Then add a second one and do a
        three-way-rename in one transaction, making sure the watcher keeps
        track.
        """
        watcher = self.runner('watch')
        watcher.setup()
        root = self.repo.path + '/__root__/'
        src = '/__source-utf8__.html'
        app = conn.app
    
        add = app.manage_addProduct['PageTemplates'].manage_addPageTemplate
        rename = app.manage_renameObject
    
        with conn.tm:
            add(id='test1', text='test1')
        watcher.step()
        assert os.path.isdir(root + 'test1')
    
        # Not sure how to apply this specifically to the secondary connection
        # and why it is only needed for the rename and not the adding, but it
        # seems to do the job
        userfolder = conn.app.acl_users
        user = userfolder.getUser('perfact').__of__(userfolder)
        newSecurityManager(None, user)
    
        with conn.tm:
            rename('test1', 'test2')
        watcher.step()
        assert os.path.isdir(root + 'test2')
        assert not os.path.isdir(root + 'test1')
    
        with conn.tm:
            add(id='test1', text='test2')
        watcher.step()
>       assert os.path.isdir(root + 'test1')
E       AssertionError: assert False
E        +  where False = <function isdir at 0x7f661bf42160>(('/tmp/tmp6tuvwh1b/__root__/' + 'test1'))
E        +    where <function isdir at 0x7f661bf42160> = <module 'posixpath' from '/usr/lib/python3.9/posixpath.py'>.isdir
E        +      where <module 'posixpath' from '/usr/lib/python3.9/posixpath.py'> = os.path

perfact/zodbsync/tests/test_sync.py:395: AssertionError
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,3 +1,6 @@
+4.1.5.dev0
+  * Fix timing issues in watcher
+
 4.1.4
   * Add subcommand `exec` to capture many git operations
 

--- a/perfact/zodbsync/commands/watch.py
+++ b/perfact/zodbsync/commands/watch.py
@@ -58,20 +58,7 @@ class Watch(SubCommand):
         effect up to this ID is visible in the current transaction and every
         effect for transaction IDs above this are not yet visible.
         '''
-        # by default, take the last transaction in the undo record
-        records = self.app._p_jar.db().undoInfo(0, 1)
-        if not len(records):
-            # after packing, no undo records exist. Return zero
-            self.last_visible_txn = b'\x00'*8
-            return
-        self.last_visible_txn = base64.b64decode(records[0]['id'])
-        # check if there are transactions we do not see yet
-        if getattr(self.app._p_jar, '_txn_time', None):
-            # if this is set, it is the first transaction that we can not yet
-            # see, so we subtract one
-            self.last_visible_txn = _decrement_txnid(
-                self.app._p_jar._txn_time
-            )
+        self.last_visible_txn = self.app._p_jar._db.lastTransaction()
 
     def _store_last_visible_txn(self):
         '''
@@ -134,12 +121,14 @@ class Watch(SubCommand):
         # * _ltid: the transaction id that was last read.
         # It is also possible to iterate over FileIterator, which yields
         # transactions in the form of a TransactionRecord
+        self.changed_oids = set()
+        if txn_start > txn_stop:
+            return
         storage = ZODB.FileStorage.FileIterator(
             self.datafs_path,
             start=txn_start,
             stop=txn_stop,
         )
-        self.changed_oids = set()
         for txn in storage:
             # Each TransactionRecord has the following fields:
             # * _pos: Position of the first data header
@@ -387,8 +376,8 @@ class Watch(SubCommand):
         # collect a list of all changed paths and record them recursively.
 
         self.acquire_lock(timeout=300)
-        self.sync.tm.begin()
         self._set_last_visible_txn()
+        self.sync.tm.begin()
         self._init_tree(self.app)
 
         # the transaction ID stored on disk is the last transaction whose
@@ -451,15 +440,15 @@ class Watch(SubCommand):
         self.acquire_lock(timeout=300)
         self.register_signals()
 
-        # make sure we see a consistent snapshot, even though we later
-        # abort this transaction since we do not write anything
-        self.sync.tm.begin()
         start_txnid = _increment_txnid(self.last_visible_txn)
         self._set_last_visible_txn()
         self._read_changed_oids(
             txn_start=start_txnid,
             txn_stop=self.last_visible_txn,
         )
+        # make sure we see a consistent snapshot, even though we later
+        # abort this transaction since we do not write anything
+        self.sync.tm.begin()
         self._update_objects()
         self.sync.tm.abort()
 

--- a/perfact/zodbsync/tests/test_sync.py
+++ b/perfact/zodbsync/tests/test_sync.py
@@ -1,4 +1,5 @@
 import os
+import time
 import os.path
 import base64
 import subprocess
@@ -415,6 +416,23 @@ class TestSync():
         self.run('playback', '/')
         assert self.app.index_html.ZCacheable_getManagerId() == "http_cache"
 
+    def watcher_step_until(self, watcher, cond):
+        """
+        After we do some changes on the secondary connection for the watcher
+        tests, the primary connection might not immediately see the change.
+        This helper function checks for a condition with several retries and
+        small waiting in between, only failing if the condition keeps being
+        false.
+        """
+        success = False
+        for i in range(5):
+            watcher.step()
+            success = cond()
+            if success:
+                break
+            time.sleep(0.5)
+        assert success
+
     def test_watch_change(self, conn):
         """
         Start the watcher, change something using the second connection without
@@ -430,8 +448,8 @@ class TestSync():
         watcher.step()
         assert 'TestRole' not in open(fname).read()
         conn.tm.commit()
-        watcher.step()
-        assert 'TestRole' in open(fname).read()
+        self.watcher_step_until(watcher,
+                                lambda: 'TestRole' in open(fname).read())
 
     def test_watch_move(self, conn):
         """
@@ -451,8 +469,8 @@ class TestSync():
 
         with conn.tm:
             add(id='test1', text='test1')
-        watcher.step()
-        assert os.path.isdir(root + 'test1')
+        self.watcher_step_until(watcher,
+                                lambda: os.path.isdir(root + 'test1'))
 
         # Not sure how to apply this specifically to the secondary connection
         # and why it is only needed for the rename and not the adding, but it
@@ -463,13 +481,13 @@ class TestSync():
 
         with conn.tm:
             rename('test1', 'test2')
-        watcher.step()
-        assert os.path.isdir(root + 'test2')
+        self.watcher_step_until(watcher, lambda: os.path.isdir(root + 'test2'))
         assert not os.path.isdir(root + 'test1')
 
         with conn.tm:
             add(id='test1', text='test2')
-        watcher.step()
+        self.watcher_step_until(watcher, lambda: os.path.isdir(root + 'test1'))
+
         assert os.path.isdir(root + 'test1')
         assert open(root + 'test1' + src).read() == 'test2'
         assert open(root + 'test2' + src).read() == 'test1'
@@ -478,7 +496,10 @@ class TestSync():
             rename('test1', 'tmp')
             rename('test2', 'test1')
             rename('tmp', 'test2')
-        watcher.step()
+        self.watcher_step_until(
+            watcher,
+            lambda: open(root + 'test1' + src).read() == 'test1',
+        )
         assert open(root + 'test1' + src).read() == 'test1'
         assert open(root + 'test2' + src).read() == 'test2'
 

--- a/setup.py
+++ b/setup.py
@@ -12,7 +12,7 @@ else:
 
 setuptools.setup(
     name='perfact-zodbsync',
-    version='4.1.4',
+    version='4.1.5.dev0',
     description='Zope Recorder and Playback',
     long_description=''' ''',
     author='Ján Jockusch et.al.',

